### PR TITLE
Refactor Tensor::to to call a primitive that is not copy_.

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61458

Context
-------
functorch is unable to vmap(grad(f)) when f contains a .to
call. This is because .to (when it is not a no-op) decomposes
to .copy_ under grad and the .copy_ is not compatible with vmap.

Fix
---
The fix for this is to have all Tensor::to variants call a new operator,
`_to_copy`, that always copies and is a primitive w.r.t. autograd so
that autograd decomposes Tensor::to into a call to `_to_copy`.
(This is related to https://github.com/pytorch/pytorch/issues/60956,
please let me know if you want to bikeshed the naming).

In order to get this done I had to do a bit of refactoring. All of the
`::to` implementations now call `to_impl` which may call `_to_copy`.

Autograd codegen changes
------------------------

The second thing I had to do was modify the autograd codegen. Right now,
autograd assumes that every output is either statically known to be
differentiable or not differentiable at codegen time. `_to_copy` is a
little special because its differentiability depends on the output
dtype. e.g. `torch.randn(3, requires_grad=True).to(torch.long)` is non
differentiable. To get this to work:
- I changed how `output_differentiability` in derivatives.yaml work.
- output_differentiability can now accept "conditions" for each of the
output arguments. A "condition" is some C++ code.
- We currently only support `output_differentiability` with conditions
if there is a single output. This is for convenience and can be changed
in the future.
- I added a new `output_differentiability_conditions` field to
DifferentiabilityInfo. This gets populated in load_derivatives.yaml
- forward-mode and reverse-mode AD take
`output_differentiability_conditions` into account.

Here's how the generated code for `VariableType::_to_copy`
[looks
like](https://gist.github.com/zou3519/93462df4bda1837acee345205b7cc849)
No other autogenerated code gets modified by this PR.

Performance benchmarking
------------------------
- I benchmarked [three
cases that demonstrate overhead](https://gist.github.com/zou3519/5b6985e6906b80eec5a0dd94ed5b6a1a).
- Case A: No-op .to(). Instruction count went from 50223 to 25623. I
have no clue why but this is a good thing.
- Case B: not-no-op .to(). Instruction count went from 665291 to 671961.
This is expected; `_to_copy` adds an additional dispatch.
- Case C: not-no-op .to() forward pass and backward pass. Instruction count
went from 4022841 to 4030057. This PR adds
an additional dispatch to .to() (so there should be one additional
dispatch in the forward pass) so this number looks reasonable.

Test Plan
---------
- test_torch.py has a test_to
- test_cuda.py has test_to*
- test_autograd has tests (test_type_conversions) that exercise the
reverse-mode path
- test_ops.py has some tests (like log_softmax) that exercise the
reverse-mode and forward-mode AD path.
- test_quantization, test_namedtensor all exercise tensor.to as well.

Differential Revision: [D29801652](https://our.internmc.facebook.com/intern/diff/D29801652)

### CODE REVIEW

**Comment by zou3519** [comment]
NB: not ready for review yet, I just want CI signal

**Comment by zou3519** [comment]
Some interesting interaction with https://github.com/pytorch/pytorch/blob/0263865bfe672eef0c71288b9ec886978a2f208a/torch/csrc/autograd/functions/utils.h#L65 :
```
x = torch.randn(3, requires_grad=True)
x.long()
```

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29801652).

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29801652).

**Comment by zou3519** [comment]
Putting the codeowners back; this is ready for review now. Thanks in advance

File: aten/src/ATen/native/TensorConversions.cpp [hunk 1]
```diff
@@ -61,6 +76,39 @@ static inline Tensor to_impl(const Tensor& self, const TensorOptions& options, b
...
+    bool copy,
+    c10::optional<c10::MemoryFormat> optional_memory_format) {
+  auto memory_format = optional_memory_format.value_or(MemoryFormat::Preserve);
+
+  // fast path
+  if (is_null_or_equal_to(dtype, self.dtype().toScalarType()) &&
+      is_null_or_equal_to(layout, self.layout()) &&
+      is_null_or_equal_to(device, self.device()) &&
+      !copy &&
+      (memory_format == MemoryFormat::Preserve ||
```
Comment (albanD) [inline aten/src/ATen/native/TensorConversions.cpp:103]: Why not `is_null_or_equal_to(optional_memory_format, MemoryFormat::Preserve)` ?
Also I'm not sure what is the suggest case below doing?

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -383,6 +383,11 @@
   self: grad
   result: auto_linear
 
+- name: _to_copy(Tensor self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, bool non_blocking=False, MemoryFormat? memory_format=None) -> Tensor
+  self: _to_copy_backward(grad, self.options())
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:387]: Can't this one be inline? Isn't there a to() overload that takes options as input?

File: tools/autograd/gen_python_functions.py [hunk 3]
```diff
@@ -84,6 +84,7 @@
     '_cholesky.*', '_triangular_solve.*', '_qr.*', '_symeig.*', '_svd.*',
     'slice', 'randint(_out)?',
     'item', '_local_scalar_dense', 'to',
+    '_to_copy',
```
Comment (albanD) [inline tools/autograd/gen_python_functions.py:87]: Is that necessary?
You can use the python version to test it if needed?

File: tools/autograd/gen_variable_type.py [hunk 4]
```diff
@@ -248,7 +249,7 @@
 
 SETUP_ANY_REQUIRES_GRAD = CodeTemplate("""\
 auto _any_requires_grad = compute_requires_grad( ${args_with_derivatives} );
-(void)_any_requires_grad;
+${extra_differentiability_conditions}(void)_any_requires_grad;
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:252]: missing new line?

File: tools/autograd/gen_variable_type.py [hunk 5]
```diff
@@ -826,6 +833,10 @@ def emit_fw_derivatives() -> List[str]:
                                        'formula has been defined for it. This case should only happen for function that '
                                        'take a single TensorList as input. All other cases are not supported right now.')
                 requires_fw_grad = "true"
+            if fn.info and fn.info.output_differentiability_conditions:
+                assert len(fn.info.output_differentiability_conditions) == 1
+                requires_fw_grad = \
+                    f'({requires_fw_grad}) && ({fn.info.output_differentiability_conditions[0]})'
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:839]: I think you want the extra condition first. Because you don't want to try and access forward grads for Tensors that you shouldn't.

File: tools/autograd/load_derivatives.py [hunk 6]
```diff
@@ -378,6 +379,17 @@ def set_up_derivatives(f: NativeFunction) -> Tuple[
     # NB: Removes 'output_differentiability' from defn dictionary
     #     `None` means all differentiable.
     output_differentiability = defn.pop('output_differentiability', None)
+    output_differentiability_conditions = None
+    if output_differentiability and any([isinstance(diff, str) for diff in output_differentiability]):
+        if len(output_differentiability) != 1:
+            raise RuntimeError(f'Not supported: for {specification} output_differentiability '
+                               f'must either '
```
Comment (albanD) [inline tools/autograd/load_derivatives.py:386]: nit: very short line?

File: tools/autograd/load_derivatives.py [hunk 7]
```diff
@@ -378,6 +379,17 @@ def set_up_derivatives(f: NativeFunction) -> Tuple[
...
+                               f'be List[bool] or a List[str] where each str is a '
+                               f'condition. In the case where it is a condition, '
+                               f'we only support single-output functions. '
+                               f'Please file us an issue. ')
+        output_differentiability_conditions = [output_differentiability[0]]
```
Comment (albanD) [inline tools/autograd/load_derivatives.py:391]: You unpack a list of one element to put it in a list with one element? :)

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 8]
```diff
@@ -175,6 +175,17 @@ std::tuple<Tensor, Tensor> _euclidean_dist_backward(const Tensor & grad, const T
...
 }
 
+Tensor _to_copy_backward(const Tensor& grad, const c10::TensorOptions& input_options) {
+  return _to_copy(
+      grad,
+      input_options.dtype().toScalarType(),
+      input_options.layout(),
+      input_options.device(),
+      input_options.pinned_memory(),
+      /*non_blocking=*/false,
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:185]: Why isn't the non-blocking propagated here?

File: aten/src/ATen/native/TensorConversions.cpp [hunk 9]
```diff
@@ -71,42 +119,52 @@ Tensor to(
...
-  if (options.has_device()) {
-    options = options.device(ensure_has_index(options.device()));
-  }
-  auto specified_options = self.options().merge_in(options);
-  return to_impl(self, specified_options, non_blocking, copy);
+  return to_impl(
+      self,
+      dtype,
+      layout,
+      device.has_value() ? ensure_has_index(*device) : device,
```
Comment (albanD) [inline aten/src/ATen/native/TensorConversions.cpp:126]: nit: make ensure_has_index overload for optional?

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (zou3519) [reply tools/autograd/derivatives.yaml:387]: `_to_copy_backward` calls `_to_copy` instead of `to()` for efficiency. There is no overload of `_to_copy` that takes options as input. `_to_copy` with arguments is a bit long so that's why this isn't inline

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 8 above)
Comment (zou3519) [reply torch/csrc/autograd/FunctionsManual.cpp:185]: It's not propagated in copy backward: https://github.com/pytorch/pytorch/blob/48af9de92ffae89618a22291738189e3110d14ed/torch/csrc/autograd/functions/tensor.cpp#L34. Since _to_copy is replacing copy backward for `at::to`, I wanted to replicate the behavior.

Another way to think about it is: It doesn't matter in the backward pass if `_to_copy` blocks or not; it only matters to the user that it blocks in the forward pass.

File: tools/autograd/gen_variable_type.py (hunk 4 above)
Comment (zou3519) [reply tools/autograd/gen_variable_type.py:252]: The problem with adding a new line is that if `${extra_differentiability_conditions}` is '' (which it is most of the time), then there is an extra new line in all the generated code. Maybe that's not a bad thing, what are your thoughts?

File: tools/autograd/gen_python_functions.py (hunk 3 above)
Comment (zou3519) [reply tools/autograd/gen_python_functions.py:87]: The python codegen chokes on the TensorOptions-like args so this is actually necessary

File: aten/src/ATen/native/TensorConversions.cpp (hunk 9 above)
Comment (zou3519) [reply aten/src/ATen/native/TensorConversions.cpp:126]: Aha, didn't realize it had one. Thanks!

File: aten/src/ATen/native/TensorConversions.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/TensorConversions.cpp:103]: > Why not is_null_or_equal_to(optional_memory_format, MemoryFormat::Preserve)

No reason, I can change it

> Also I'm not sure what is the suggest case below doing?

I copied this from the original `to_impl` code, but here's the explanation:
- The suggest case is saying that "if self has the same memory_format as the one passed to `to_impl`, then we don't need to do anything"
- Unfortunately we can't just query `self.options().memory_format()` and compare it to `optional_memory_format` because of a quirk of how MemoryFormat is implemented (see next two bullet points)
- Sometimes a ChannelsLast tensor doesn't actually get marked as being ChannelsLast.
- In order to tell if a ChannelsLast tensor is actually ChannelsLast, we have to do `self.suggest_memory_format(memory_format)` or `self.is_contiguous(memory_format=memory_format)` and this code is doing the former

Comment (albanD) [reply aten/src/ATen/native/TensorConversions.cpp:103]: ok!
Thanks for the full explanation! Is that captured somewhere in a comment in code?

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (albanD) [reply tools/autograd/derivatives.yaml:387]: > _to_copy instead of to() for efficiency

Is the difference really that big?
Also maybe the gradient already has the right properties and you don't want to copy it?

File: tools/autograd/gen_python_functions.py (hunk 3 above)
Comment (albanD) [reply tools/autograd/gen_python_functions.py:87]: Ho ok!

File: tools/autograd/gen_variable_type.py (hunk 4 above)
Comment (albanD) [reply tools/autograd/gen_variable_type.py:252]: Ho right. I would say it is not the end of the world to have the extra empty line in the generated code. But it will look weird indeed.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 8 above)
Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:185]: For CPU -> CUDA it doesn't matter indeed. Because the copy the other way around cannot be non-blocking.
But for CUDA -> CUDA copy maybe we should do that actually?

cc @mcarilli ?

File: aten/src/ATen/native/TensorConversions.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/TensorConversions.cpp:103]: I don't think so, I'll put a comment in about this over here

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (zou3519) [reply tools/autograd/derivatives.yaml:387]: > Is the difference really that big?

The difference is another redispatch, so it's really not that big.

> Also maybe the gradient already has the right properties and you don't want to copy it?

This can't happen if the user calls `tensor.to(...)`, but it could happen if a user uses `_to_copy` directly. In the future these `_copy` APIs may be exposed to users, so let's replace this with `::to`

Comment (albanD) [reply tools/autograd/derivatives.yaml:387]: > This can't happen if the user calls tensor.to(...)

Really? I could see a case where we clone in the forward to satisfy the requires layout, but the layout is already correct in the backward. no?

Comment (zou3519) [reply tools/autograd/derivatives.yaml:387]: Good point, the layout of gradients is not always the same as the tensor (though we try to make it so and there are some special cases about leaf nodes and their .grad fields)

File: aten/src/ATen/native/TensorConversions.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/TensorConversions.cpp:103]: Reading this code more... the suggest case seems like a bug: https://github.com/pytorch/pytorch/issues/62027 .
Since the code as-is reproduces the behavior before the refactor, is it fine if I leave it as-is for now and then fix it whenever #62027 gets a resolution?

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29801652).

**Review by albanD — APPROVED** [review]
LGTM
Thanks for the fix!

File: aten/src/ATen/native/TensorConversions.cpp [hunk 10]
```diff
@@ -61,6 +83,45 @@ static inline Tensor to_impl(const Tensor& self, const TensorOptions& options, b
...
+      // for non-contiguous tensors that don't look ChannelsLast.
+      // This leads to `x.to(memory_format=torch.contiguous_format)`
+      // not always returning a contiguous tensor. Maybe the condition should
+      // actually be `self.is_contiguous(memory_format)`?
+      (memory_format == MemoryFormat::Preserve ||
+       self.suggest_memory_format() == memory_format)) {
+    return self;
+  }
+
+  return at::_to_copy(
```
Comment (ngimel) [inline aten/src/ATen/native/TensorConversions.cpp:121]: does it increase the number of dispatches that `to` goes through? Before it used to be `to` and `copy_`, now it's `to`, `_to_copy` and `copy_`?

**Comment by malfet** [comment]
This PR broke OSS CI which was clearly indicated by following failure on PR: https://app.circleci.com/pipelines/github/pytorch/pytorch/355891/workflows/376515a2-05d4-44bc-9be6-6240d3509021/jobs/15010695

File: aten/src/ATen/native/TensorConversions.cpp (hunk 10 above)
Comment (zou3519) [reply aten/src/ATen/native/TensorConversions.cpp:121]: Yes, that is correct. There are benchmarks in the PR body that demonstrate this. If we care about overhead, then in the future we can make `_to_copy` directly call TensorIterator instead of calling `copy_`.
